idx/memory/memory.go

```diff
@@ -465,6 +467,228 @@ func (m *MemoryIdx) TagDetails(orgId int, key, filter string, from int64) (map[s
 	return res, nil
 }
 
+// AutoCompleteTags is used to generate a list of possible values that could
+// complete a given prefix. It also accepts additional conditions to further
+// narrow down the result set.
+//
+// tagPrefix:   the string to be completed
+// expressions: tagdb expressions in the same format as graphite uses
+// from:        only tags will be returned that have at least one metric
+//              with a LastUpdate above from
+// limit:       the maximum number of results to return, the results will
+//              always be sorted alphabetically for consistency between
+//              consecutive queries and the limit is applied after sorting
+//
+func (m *MemoryIdx) AutoCompleteTags(orgId int, tagPrefix string, expressions []string, from int64, limit uint) ([]string, error) {
+	res := make([]string, 0)
+
+	// only if expressions are specified we need to build a tag query.
+	// otherwise, the generation of the result set is much simpler
+	if len(expressions) > 0 {
+		// incorporate the tag prefix into the tag query expressions
+		if len(tagPrefix) > 0 {
+			expressions = append(expressions, fmt.Sprintf("__tag^=%s", tagPrefix))
+		}
+
+		query, err := NewTagQuery(expressions, from)
+		if err != nil {
+			return nil, err
+		}
+
+		// only acquire lock after we're sure the query is valid
+		m.RLock()
+		defer m.RUnlock()
+
+		tags, ok := m.tags[orgId]
+		if !ok {
+			return nil, nil
+		}
+
+		resMap := query.RunGetTags(tags, m.DefById)
+		for tag := range resMap {
+			res = append(res, tag)
+		}
+
+		sort.Strings(res)
+		if uint(len(res)) > limit {
+			res = res[:limit]
+		}
+	} else {
+		m.RLock()
+		defer m.RUnlock()
+
+		tags, ok := m.tags[orgId]
+		if !ok {
+			return nil, nil
+		}
+
+		tagsSorted := make([]string, 0, len(tags))
+		for tag := range tags {
+			if len(tagPrefix) > 0 && (len(tagPrefix) > len(tag) || tag[:len(tagPrefix)] != tagPrefix) {
```

Comment:
I think we can replace this entire condition with  `if !strings.HasPrefix(tag, tagPrefix) {`
tagPrefix being "" doesn't need anything special, it just works.

Reply:
yep, that works :+1: